### PR TITLE
Openstack selfmanaged secrets

### PR DESCRIPTION
**Pull request by pierreprinetti**
With this change, the `ipi-conf-openstack` job will import a clouds.yaml
based on which "CLUSTER_TYPE" is defined in the job configuration.

Also with this change, the OpenStack step-registry jobs use
self-maintained secrets for storing cloud credentials, and only mounts
them in the step where they are required.

Note that contrary to https://github.com/openshift/release/pull/14516, **this patch does NOT try to solve how OpenStack flavor names (and external network name) are injected in the jobs**: the cloud-specific configuration is still injected as an environment variable in the workflow definition. As a result:
*  It is only possible to migrate ALL `e2e-openstack-ipi` jobs to a specific cloud (and not, for example, "only 4.8" or "only master"). same applies to the `csi` and `manifests` jobs;
* To migrate one job to another cloud, first change the `cluster_profile` in the configuration, then set the appropriate flavor and network names in the workflow definition's `env` section.

/cc @mandre @adduarte

In separate commits:
* Quote variables to avoid globbing
* Set SHARED_DIR as readonly where possible

### CODE REVIEW

File: ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-ref.yaml [hunk 1]
```diff
@@ -7,10 +7,7 @@ ref:
       cpu: 10m
       memory: 100Mi
   env:
-    - name: OS_CLOUD
-      documentation: Name of cloud to use from ${CLUSTER_PROFILE_DIR}/clouds.yaml file
-    - name: OPENSTACK_EXTERNAL_NETWORK
-      documentation: Name of external network to attach FIPs to.
```
Comment (mandre) [inline ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-ref.yaml:13]: we should probably keep the variable documentation.

**Review by mandre — COMMENTED** [review]
This looks good, thanks.

File: ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-ref.yaml (hunk 1 above)
Comment (pierreprinetti) [reply ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-ref.yaml:13]: woops! thanks

File: ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-ref.yaml [hunk 2]
```diff
@@ -6,16 +6,18 @@ ref:
...
-    - name: OPENSTACK_EXTERNAL_NETWORK
-    - name: OPENSTACK_COMPUTE_FLAVOR
-    - name: OS_CLOUD
-    - name: BASE_DOMAIN
-    - name: KURYR_ENABLED
-      default: "false" 
+  - name: OS_CLIENT_CONFIG_FILE_VEXXHOST
+    default: '/var/run/cluster-secrets-openstack-vexxhost/clouds.yaml'
+  - name: OS_CLIENT_CONFIG_FILE_MOC
+    default: '/var/run/cluster-secrets-openstack-moc/clouds.yaml'
```
Comment (adduarte) [inline ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-ref.yaml:20]: can't we merge this two into one file?

File: ci-operator/step-registry/openshift/e2e/openstack/ipi/openshift-e2e-openstack-ipi-workflow.yaml [hunk 3]
```diff
@@ -9,8 +9,7 @@ workflow:
     - chain: ipi-openstack-post
     env:
       OPENSTACK_EXTERNAL_NETWORK: "external"
-      OS_CLOUD: "openstack-cloud"
-      OPENSTACK_COMPUTE_FLAVOR: "m1.s2.xlarge"
+      OPENSTACK_INSTANCE_FLAVOR: "m1.s2.xlarge"
```
Comment (adduarte) [inline ci-operator/step-registry/openshift/e2e/openstack/ipi/openshift-e2e-openstack-ipi-workflow.yaml:12]: :) This chances should really be in a different PR, but its not that big of a deal. There is actually not technical reason to change the name of the variable just because the name is strange.

File: ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh [hunk 4]
```diff
@@ -34,7 +32,7 @@ else
     exit 1
 fi
 
-# Lets  check the syntax of yaml file by reading it.
+# Check the syntax of yaml file by reading it.
```
Comment (adduarte) [inline ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh:35]: again why?  pr changes should really be only about what is necessary not what we think it should be based on word preference. This is absolutely superfluous

File: ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-commands.sh [hunk 5]
```diff
@@ -4,13 +4,31 @@ set -o nounset
...
+    OS_CLIENT_CONFIG_FILE="$OS_CLIENT_CONFIG_FILE_VEXXHOST"
+    ;;
+  openstack)
+    OS_CLIENT_CONFIG_FILE="$OS_CLIENT_CONFIG_FILE_MOC"
+    ;;
+  *)
+    echo >&2 "Unsupported cluster type '${CLUSTER_TYPE}'"
+    exit 1
+    ;;
+esac
```
Comment (adduarte) [inline ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-commands.sh:31]: if you combine /var/run/cluster-secrets-openstack-vexxhost and /var/run/cluster-secrets-openstack-moc into a single file this becomes unecessary.

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh [hunk 6]
```diff
@@ -23,8 +23,9 @@ azure4) export AZURE_AUTH_LOCATION=${CLUSTER_PROFILE_DIR}/osServicePrincipal.jso
 gcp) export GOOGLE_CLOUD_KEYFILE_JSON=${CLUSTER_PROFILE_DIR}/gce.json;;
 kubevirt) export KUBEVIRT_KUBECONFIG=${HOME}/.kube/config;;
 vsphere) ;;
-openstack) export OS_CLIENT_CONFIG_FILE=${CLUSTER_PROFILE_DIR}/clouds.yaml ;;
-openstack-vexxhost) export OS_CLIENT_CONFIG_FILE=${CLUSTER_PROFILE_DIR}/clouds.yaml ;;
+openstack|openstack-vexxhost)
+  export OS_CLIENT_CONFIG_FILE="${SHARED_DIR}/clouds.yaml"
+  ;;
```
Comment (adduarte) [inline ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:28]: This is awsome!

File: ci-operator/step-registry/openstack/deprovision/deletefips/openstack-deprovision-deletefips-ref.yaml [hunk 7]
```diff
@@ -6,9 +6,6 @@ ref:
     requests:
       cpu: 1000m
       memory: 100Mi
-  env:
-    - name: OS_CLOUD
-      documentation: name of cloud to use from clouds.yaml
```
Comment (adduarte) [inline ci-operator/step-registry/openstack/deprovision/deletefips/openstack-deprovision-deletefips-ref.yaml:11]: leave this as is, and instead give it a default.

File: ci-operator/step-registry/openstack/test/manifests/openstack-test-manifests-commands.sh [hunk 8]
```diff
@@ -2,7 +2,8 @@
 
 set -Eeuo pipefail
 
-export OS_CLIENT_CONFIG_FILE="${CLUSTER_PROFILE_DIR}/clouds.yaml"
+export OS_CLIENT_CONFIG_FILE="${SHARED_DIR}/clouds.yaml"
+OS_CLOUD='openstack'
```
Comment (adduarte) [inline ci-operator/step-registry/openstack/test/manifests/openstack-test-manifests-commands.sh:6]: same thing, don't hardocode this in the sh script part of the file.

**Review by adduarte — COMMENTED** [review]
this is looking good. I am not going to opose this merge, but I do think there are things that could be done better.

File: ci-operator/step-registry/openstack/deprovision/deletefips/openstack-deprovision-deletefips-ref.yaml (hunk 7 above)
Comment (adduarte) [reply ci-operator/step-registry/openstack/deprovision/deletefips/openstack-deprovision-deletefips-ref.yaml:11]: if instead you do this: 
env: 
   - name: OS_CLOUD
    documentation: name of the cloud provider to use from clouds.yaml
    default: "opesntack"

You accomplish in  what you want to do with one single line addition. , the next change is unecessary, and the step stays general.

**Comment by adduarte** [comment]
This looks good to me. 
lets get a green test. 
we can submit follow ups if necessary

**Comment by pierreprinetti** [comment]
@adduarte
As sorted out offline, I am closing this PR in favor of your proposal.

File: ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-ref.yaml (hunk 2 above)
Comment (mandre) [reply ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-ref.yaml:20]: I see two problems with merging the secrets together:
- if the test container somehow gets compromises, that exposes the secret for all cloud platforms
- the clouds.yaml to environment variable conversion currently expects only one entry in the clouds.yaml file, see https://bugzilla.redhat.com/show_bug.cgi?id=1918680. Not that this is a big issue, since I'm going to make it more robust in https://issues.redhat.com/browse/OSASINFRA-2313.
